simpleline/event_loop/main_loop.py

```diff
@@ -70,13 +77,47 @@ def run(self):
         if self._quit_callback:
             self._quit_callback()
 
+    def execute_new_loop(self, signal):
+        """Starts the new event loop and pass `signal` in it.
+
+        This is required for processing a modal screens.
+
+        :param signal: signal passed to the new event loop
+        :type signal: `AbstractSignal` based class
+        """
+        self._active_queue = EventQueue()
+        self._event_queues.append(self._active_queue)
+        self.enqueue_signal(signal)
+        self._mainloop()
+
+    def close_loop(self):
+        """Close active event loop.
+
+        Close an event loop created by the `execute_new_loop()` method.
+        """
+        self.process_signals()
+        self._event_queues.pop()
+        try:
+            self._active_queue = self._event_queues[-1]
+        except IndexError:
+            raise ExitMainLoop()
+
+        raise ExitMainLoop()
+
     def enqueue_signal(self, signal):
         """Enqueue new event for processing.
 
+        Enqueue signal to the most inner queue (nearest to the active queue) where the `signal.source` belongs.
+        If it is not belong anywhere enqueue it to the active one.
+
         :param signal: event which you want to add to the event queue for processing
         :type signal: instance based on AbstractEvent class
         """
-        self._queue_instance.put(signal)
+        for i in range(len(self._event_queues), 0):
+            queue = self._event_queues[i]
+            if queue.enqueue_if_source_belongs(signal, signal.source):
+                return
+        self._active_queue.enqueue(signal)
```

Comment:
Could be just `for queue in reversed(self._event_queues): ...`
https://docs.python.org/3/library/functions.html#reversed